### PR TITLE
Fixed Coordinate Encodings in TEX and TEXS instructions

### PR DESCRIPTION
**Pull request by FernandoS27**
After intense RE and some docs on nouveau, I got all the encodings for coordinates, this fixes a bunch of graphical issues. Among them, the famous fish-eye effect on SMO.

### CODE REVIEW

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 1]
```diff
@@ -2643,37 +2669,63 @@ class GLSLGenerator {
                 const bool depth_compare =
                     instr.texs.UsesMiscMode(Tegra::Shader::TextureMiscMode::DC);
                 u32 num_coordinates = TextureCoordinates(texture_type);
-                if (depth_compare)
-                    num_coordinates += 1;
+                auto process_mode = instr.texs.GetTextureProcessMode();
```
Comment (ReinUsesLisp) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: `const auto`

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 2]
```diff
@@ -2491,61 +2491,81 @@ class GLSLGenerator {
...
+                std::string lod_value;
+                if (process_mode != Tegra::Shader::TextureProcessMode::LZ &&
+                    process_mode != Tegra::Shader::TextureProcessMode::None) {
+                    start_index_b = 1;
+                    lod_value = regs.GetRegisterAsFloat(instr.gpr20);
+                }
+
+                std::string depth_value;
+                if (depth_compare) {
+                    depth_value = regs.GetRegisterAsFloat(instr.gpr20.Value() + start_index_b);
```
Comment (ReinUsesLisp) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: By using a C++ temporal variable, register overwrites may occur. Use a GLSL variable instead.

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 3]
```diff
@@ -2643,37 +2669,63 @@ class GLSLGenerator {
...
                     instr.texs.UsesMiscMode(Tegra::Shader::TextureMiscMode::DC);
                 u32 num_coordinates = TextureCoordinates(texture_type);
-                if (depth_compare)
-                    num_coordinates += 1;
+                auto process_mode = instr.texs.GetTextureProcessMode();
+                std::string lod_value;
+                u32 lod_offset = 0;
+                if (process_mode == Tegra::Shader::TextureProcessMode::LL) {
+                    if (num_coordinates > 2) {
+                        lod_value = regs.GetRegisterAsFloat(instr.gpr20.Value() + 1);
```
Comment (ReinUsesLisp) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Ditto about temporal variables.

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 4]
```diff
@@ -2643,37 +2669,63 @@ class GLSLGenerator {
...
+                        lod_offset = 2;
+                    } else {
+                        lod_value = regs.GetRegisterAsFloat(instr.gpr20);
```
Comment (ReinUsesLisp) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: Ditto about temporal variables.

File: src/video_core/renderer_opengl/gl_shader_decompiler.cpp [hunk 5]
```diff
@@ -2643,37 +2669,63 @@ class GLSLGenerator {
...
+                        lod_offset = 1;
+                    }
+                }
 
                 switch (num_coordinates) {
+                case 1: {
+                    const std::string x = regs.GetRegisterAsFloat(instr.gpr8);
+                    coord = "float coords = " + x + ';';
```
Comment (ReinUsesLisp) [inline src/video_core/renderer_opengl/gl_shader_decompiler.cpp]: IMO you can avoid declaring `x` here.

**Comment by FernandoS27** [comment]
@ReinUsesLisp Went off ahead and fixed the texture overwriting through an easier method without affecting the inputs.

**Comment by bunnei** [comment]
@ReinUsesLisp please re-review
@FernandoS27 needs a rebase

**Pull request merged**